src/coreclr/src/jit/codegenlinear.cpp

```diff
@@ -741,15 +730,73 @@ void CodeGen::genCodeForBBlist()
 #endif // !FEATURE_EH_FUNCLETS
 
             case BBJ_NONE:
-            case BBJ_COND:
             case BBJ_SWITCH:
                 break;
 
+            case BBJ_ALWAYS:
+                inst_JMP(EJ_jmp, block->bbJumpDest);
+                __fallthrough;
+            case BBJ_COND:
+                if (block->bbJumpDest->bbFlags & BBF_FIRST_BLOCK_IN_INNERLOOP)
+                {
+                    // Track the destination IG which is the first block of inner loop.
+                    // In emitter, this will be used to calculate total instructions present
+                    // in all IGs that participate in a loop.
+
+                    insGroup* srcIG = GetEmitter()->emitCurIG;
+                    insGroup* dstIG = (insGroup*)block->bbJumpDest->bbEmitCookie;
+
+                    // Only track back edges to the loop.
+                    // Here dstIG != nullptr checks if we have already generated dstIG for a block.
+                    // If block->bbJumpDest was a forward block, it might have not been created yet.
+                    // We don't rely on (block->bbJumpDest->bbNum <= block->bbNum) because the basic
+                    // block numbering is not guaranteed to be sequential.
+                    if (dstIG != nullptr && dstIG->igNum <= srcIG->igNum)
+                    {
+                        srcIG->igLoopBackEdge = dstIG;
+#ifdef DEBUG
+                        if (verbose)
+                        {
+                            printf("** IG_%d jumps back to IG_%d forming a loop.\n", srcIG->igNum, dstIG->igNum);
+                        }
+#endif
+                    }
+                }
+                break;
+
             default:
                 noway_assert(!"Unexpected bbJumpKind");
                 break;
         }
 
+#if defined(TARGET_XARCH)
```

Comment:
Currently we will only do for xarch. ArmArch support will be added (as mentioned in https://github.com/dotnet/runtime/issues/43227) if we see real benefits because of loop alignment in our perf runs.